### PR TITLE
Dompdf ^2.0 upgrade

### LINKED ISSUE #382: Upgrade Dompdf to version 2.0.0
Dompdf released the [new version 2.0.0](https://github.com/dompdf/dompdf/releases/tag/v2.0.0) a few minutes ago, we should start testing this because brings security fixes but also breaking changes.

### PR DESCRIPTION
**Pull request by alexmigf**
closes #382

Needs testing from @YordanSoares

### CODE REVIEW

**Comment by Spreeuw** [comment]
@alexmigf to be clear: this is still in draft because it's not ready for review by @YordanSoares , correct?
Some questions that came to mind when I saw the dompdf changelog:

1. We have a setting "Use alternative HTML5 parser to parse HTML" is that setting (or its behavior) affected by the new HTML parser?
2. I guess migration of the font cache data to the new JSON format is required (as part of the upgrade routines), as well as our `Font_Synchronizer` class, correct?
3. Since dompdf has breaking changes, should we also bump our plugin version to 3.0?

**Comment by alexmigf** [comment]
@Spreeuw no, I set as draft because is waiting testing not review. But I can change that if you want.

**Comment by alexmigf** [comment]
@Spreeuw 

1) Seems to be working fine.
2) I guess so.
3) I think we should.

**Comment by Spreeuw** [comment]
Regarding the HTML parser, it seems the `isHtml5ParserEnabled` option is no longer used at all though, so the setting wouldn't have any effect?

This is the current code:
https://github.com/dompdf/dompdf/blob/165ad28f0291d7ba257b236caaa4b442a28d2150/src/Dompdf.php#L473-L489

And the old code where it used the option:
https://github.com/dompdf/dompdf/blob/5f52622e79202dd469fdd3053f141d30e0901808/src/Dompdf.php#L505-L511

I no longer find any code using the `isHtml5ParserEnabled` option in dompdf

**Comment by Spreeuw** [comment]
That's just the setters and the getters in the `Options` object, not being actually used, is it?

**Comment by alexmigf** [comment]
No, doesn't seem to be used.

**Comment by alexmigf** [comment]
They are using this library now: https://github.com/Masterminds/html5-php

**Comment by Spreeuw** [comment]
> They are using this library now: https://github.com/Masterminds/html5-php

That replaces `DOMDocument` from the old code ([source](https://github.com/dompdf/dompdf/blob/5f52622e79202dd469fdd3053f141d30e0901808/src/Dompdf.php#L512-L518)). The alternative parser has been removed completely. So we should also remove the option.

**Comment by alexmigf** [comment]
## Test 1

1. Using `main` branch, check the `uploads/wpo_wcpdf_*/fonts` directory for:
    1.1 PHP cache file is present
    1.2 Plugin fonts are present
2. Change to this PR branch and simulate the plugin upgrade to version `3.0.0`:
    2.1 Legacy PHP cache file is deleted :heavy_check_mark: 
    2.2 New JSON cache file is created with only the plugin fonts :heavy_check_mark:

## test 2

1. Using this PR branch only, simulating a new install:
    1.1 New JSON cache file is created with the plugin fonts
    1.2 Plugin fonts are copied

File: includes/class-wcpdf-font-synchronizer.php [hunk 1]
```diff
@@ -72,11 +72,8 @@ public function sync( $destination, $merge_with_local = true ) {
 
 		// normalize one last time
 		$local_fonts = $this->normalize_font_paths( $local_fonts );
-
 		// rebuild font cache file
-		$fonts_export = var_export( $local_fonts, true );
-		$fonts_export = str_replace( '\'' . untrailingslashit( $destination ) , '$fontDir . \'', $fonts_export );
-		$cacheData = sprintf( "<?php return %s;%s?>", $fonts_export, PHP_EOL );
+		$cacheData   = json_encode( $local_fonts );
```
Comment (Spreeuw) [inline includes/class-wcpdf-font-synchronizer.php:76]: dompdf uses `JSON_PRETTY_PRINT`, which I think is nice to copy here too!

File: includes/class-wcpdf-font-synchronizer.php [hunk 2]
```diff
@@ -17,7 +17,7 @@ class Font_Synchronizer {
 	 *
 	 * @var string
 	 */
-	public $font_cache_filename = "dompdf_font_family_cache.php";
+	public $font_cache_filename = "installed-fonts.dist.json";
```
Comment (Spreeuw) [inline includes/class-wcpdf-font-synchronizer.php:20]: I think this shouldn't be `.dist`, right? See:
https://github.com/dompdf/dompdf/blob/79573d8b8a141ec8a17312515de8740eed014fa9/src/FontMetrics.php#L36

**Review by Spreeuw — CHANGES_REQUESTED** [review]
In your test you write this:

> 2.2 New JSON cache file is created with only the plugin fonts ✔️

But should this include the local (user installed) fonts too? Have you confirmed that the user installed fonts are correctly loaded after the migration? I'm asking because the `.dist` part in the JSON filename seems incorrect to me, but maybe I'm wrong.

File: includes/class-wcpdf-install.php [hunk 3]
```diff
@@ -404,7 +404,18 @@ protected function upgrade( $installed_version ) {
...
+		if ( version_compare( $installed_version, '3.0.0-dev-1', '<' ) ) {
+			// removes 'HTML5 parser' setting value
+			$debug_settings = get_option( 'wpo_wcpdf_settings_debug', array() );
+			if ( ! empty( $debug_settings['use_html5_parser'] ) ) {
+				unset( $debug_settings['use_html5_parser'] );
+				update_option( 'wpo_wcpdf_settings_debug', $debug_settings );
+			}
+			// force copy fonts merge to replace the fonts cache file (Dompdf <2.0)
+			WPO_WCPDF()->main->copy_fonts( $font_path, true );
+		}
```
[the remark this answers is not in the document]
Comment (alexmigf) [reply includes/class-wcpdf-install.php:418]: You mean the cache file replacement?

[the remark this answers is not in the document]
Comment (Spreeuw) [reply includes/class-wcpdf-install.php:418]: yes, moving from JSON to PHP

[the remark this answers is not in the document]
Comment (alexmigf) [reply includes/class-wcpdf-install.php:418]: Shouldn't be the contrary? From PHP to JSON? But that is already triggered by the `copy_fonts` when `$merge_with_local` is set to `true`.

**Comment by alexmigf** [comment]
> But should this include the local (user installed) fonts too? Have you confirmed that the user installed fonts are correctly loaded after the migration? I'm asking because the .dist part in the JSON filename seems incorrect to me, but maybe I'm wrong.

Yes, you're right, shouldn't have the `dist` part. User local fonts are migrated as well.

File: includes/class-wcpdf-install.php [hunk 4]
```diff
@@ -404,7 +404,18 @@ protected function upgrade( $installed_version ) {
...
+		// 3.0.0-dev-1: remove saved option 'use_html5_parser'
+		if ( version_compare( $installed_version, '3.0.0-dev-1', '<' ) ) {
+			// removes 'HTML5 parser' setting value
+			$debug_settings = get_option( 'wpo_wcpdf_settings_debug', array() );
+			if ( ! empty( $debug_settings['use_html5_parser'] ) ) {
+				unset( $debug_settings['use_html5_parser'] );
+				update_option( 'wpo_wcpdf_settings_debug', $debug_settings );
+			}
+			// force copy fonts merge to replace the fonts cache file (Dompdf <2.0)
+			WPO_WCPDF()->main->copy_fonts( $font_path, true );
```
Comment (Spreeuw) [inline includes/class-wcpdf-install.php:417]: as you correctly pointed out in previous comments, this is already done earlier in this method:
https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/blob/de3f055345804a78a51fa043ba01333e0a00fd50/includes/class-wcpdf-install.php#L218
so we don't need to repeat it here.

**Review by Spreeuw — CHANGES_REQUESTED** [review]

**Comment by YordanSoares** [comment]
I ran the same test that @alexmigf posted [here](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/pull/383#issuecomment-1167339914), and both the updating and new installation tests seems to be working well.

However, after the updating process, both `dompdf_font_family_cache.php` and `installed-fonts.json` seem to remain in the folder. I saw that the legacy PHP cache file is deleted only if you click on the **Reinstall fonts** button.

**Comment by YordanSoares** [comment]
@alexmigf I installed the `main` branch, and then update the plugin manually with the branch from this PR.

**Comment by YordanSoares** [comment]
@alexmigf I ran another test installing the `main `branch, and then update the plugin manually (under **Plugins > Add new > Upload plugin**) with the branch from this PR, and the `dompdf_font_family_cache.php` is not deleted in this step.

Also, I realized that `installed-fonts.json` is not created just after the update, but when a PDF document is requested. I'm not sure if this is the expected behavior.

**Comment by alexmigf** [comment]
@YordanSoares did you changed the plugin version to 3.0.0?

**Comment by YordanSoares** [comment]
@alexmigf that did the trick: Sorry for the mess! When I read the instructions to run the tests, I missed the part in which you mention the version change.

**Review by YordanSoares — APPROVED** [review]

**Comment by alexmigf** [comment]
@YordanSoares I think you still need to check if aren't any changes in the PDF output for each template.

**Review by Spreeuw — APPROVED** [review]

**Pull request merged**